### ISSUE TITLE
Unit doesn't restart configuration on app fail

### ISSUE THREAD

**Issue by mancubus77** [nginx/unit#194]
I'm running official Unit docker image with startup config 
```
[root@767e2b75fd3a /]# unitd --version 
unit version: 1.6
configured as ./configure --prefix=/usr --state=/var/lib/unit --control=unix:/var/run/control.unit.sock --pid=/var/run/unit.pid --log=/var/log/unit.log --tests --openssl --modules=/usr/lib64/unit/modules --libdir=/usr/lib64 --cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fPIC' --ld-opt='-Wl,-z,relro -Wl,-z,now -pie'
```
```
{
    "applications": {
        "rrl": {
          "type": "python",
          "processes": 10,
          "path": "/var/app",
          "module": "app",
          "user": "nobody",
          "group": "nobody"
        }
    },

    "listeners": {
        "*:8083": {
            "application": "rrl"
        }
    },
  "access_log": "/var/log/unit-access.log"
}

```
 When application failed because of bug in code, unit doesn't try to restart it. I intentionally put an error in flask code and got a error message on startup
```
2018/12/05 07:54:31 [info] 8#8 "rrl" application started
2018/12/05 07:54:31 [alert] 8#8 Python failed to import module "app"
  File "/var/app/app.py", line 149
    11application.run()
                ^
SyntaxError: invalid syntax
2018/12/05 07:54:31 [notice] 1#1 process 8 exited with code 1
2018/12/05 07:54:31 [warn] 7#7 failed to start application "rrl"
2018/12/05 07:54:31 [alert] 7#7 failed to apply new conf
2018/12/05 07:54:31 [alert] 6#6 failed to apply previous configuration
```
Unit is acting as an engine, meanwhile source code on host machine. Is there any way to force nginx trey to rerun application on failure?

**Comment by VBart** [comment]
@mancubus77 
Thank you for the valuable feedback.

It's impossible to determine the moment when application is fixed. Only the developer may know when to give it another try. Currently, if Unit cannot successfully pre-fork an application process it tries to fall-back to the previous configuration (or to the default one). So, in order to make another try you have to reapply the configuration.

As a workaround, you can configure dynamic application process management: http://unit.nginx.org/configuration/#process-management and Unit will start the application at the time of request arrival.

I understand possible inconvenience here. In order to overcome that inconvenience there is a new feature planned. It's called "Configuration History". All the configuration changes (successful or not) will be put in the "history" object in Unit's control API. There will be simple commands to re-apply the last failed configuration or to jump to any of the previous configurations.

**Comment by sureshjoshi** [comment]
@VBart Was there any progress on the configuration history feature?

I've run into this same problem, and it's very unintuitive to have an "unrecoverable" failure in Docker (in that, the system doesn't try to recover) - while the health checks show everything is working.

I totally understand why it happens, it's just very strange. In my case, the error ends up chaining a few microservices (my main gateway and the one that caused the problem and anything else using Unit and depending on these services). 

What I would have expected to happen was that, if the Docker image is created with a config.json - and the health check fails (e.g. the curl PUT fails), then the container fails. Given it's occurring at container creation time, there is no reason why a failure should be allowed to propagate. Runtime failures are, of course, a completely different problem.

Actually, the problem is even a bit worse... After a config failure during initial run, all subsequent restarts or relaunches of the container don't even try again with the following message:

`/usr/local/bin/docker-entrypoint.sh: /var/lib/unit/ is not empty, skipping initial configuration...`

Basically, you have to destroy the container and force re-build it in order to get it to try again (or make a config change).

**Comment by sureshjoshi** [comment]
Also, while I'm not proposing this as a solution - something like this might be handy.
In Dockerfile:

`HEALTHCHECK CMD curl --fail -X PUT --data-binary @config.json --unix-socket /var/run/control.unit.sock http://localhost/config || exit 1`

When you do a `docker ps`, you'll see an "unhealthy" status for the container if that command failed - and there are some automatic retries/intervals/etc...

So, while not perfect - it's definitely helpful as it indicates really quick that there is a problem.

**Comment by VBart** [comment]
@sureshjoshi `docker-entrypoint.sh` is used only to create the initial image. It doesn't upload configuration each time, since Unit has its own persistent state. It's more like an initialization of the database, creating tables. After that you have to work with the database (Unit configuration API) on your own.

If you'd like to avoid any persistence, then you should clean up the Unit state directory, but it's not recommended way of working.

Unit configuration is all about runtime and in fact, even this initial failure happens in runtime. After Unit starts, it tries to apply configuration saved in the state. If it fails, then it fallbacks to the default empty config. But the configuration isn't lost: if you restart the daemon, it will try to apply the configuration from the state again.

In order to check if the configuration was applied or not, you can check the log or check the API with the GET request. But if you prevent Unit from starting in case of failed configuration, then how would you reconfigure it?

> Basically, you have to destroy the container and force re-build it in order to get it to try again (or make a config change).

It looks like you're trying to abuse `docker-entrypoint.sh` for any configuration purpose. Unit provides an HTTP runtime API for configuration and your container orchestration tool should control it directly via this API, not via putting files into `docker-entrypoint.sh` directories.

Unit configuration concept is very close to a database. In order to change any data in the database, you have to run the database daemon and then connect to it. When it fails to apply the configuration on the start, that means that the data in the database is corrupted and no longer valid, you have to connect and upload a new one or just restart the daemon and try again. 

>  Was there any progress on the configuration history feature?

It's still in TODO list. This feature is from the nice to have class of features, but we have a bunch of must have essentials features that need to be implemented first. I hope that with [the expansion of development team](https://mailman.nginx.org/pipermail/unit/2020-October/000220.html) we will have faster progress.

**Comment by tippexs** [comment]
closed due to timeout
